lib/LINZ/BdeUpload.pm

```diff
@@ -369,6 +369,58 @@ use fields qw{ cfg db repository tables tmp_base tmp fid current_dataset
 
 our $tmp_prefix='tmp_bde_upload_';
 
+sub uploadDataToTempTable
+{
+    my($self,$tablename,$tmpfile,$columns) = @_;
+
+    $self->db->_setDbMessageHandler;
+
+    DEBUG('Loading file ' . $tmpfile . ' into table ' . $tablename);
+
+    my $uploadId = $self->db->uploadId;
+    my $dbh = $self->db->_dbh;
+
+    my @quotedColumns;
+    foreach my $col (split(/\|/, $columns)) {
+        push(@quotedColumns, $dbh->quote_identifier( $col ));
+    }
+
+    #DEBUG('quoted columns are ' . join(',',@quotedColumns));
+
+    my $sql = 'SELECT bde_control._bde_RefreshLock(?)';
```

Comment:
Shouldn't be using linz_bde_uploader PgSQL private methods in here. Maybe a new public API is need, or the promotion of the private API to public? If you do this then the method should be automatically available as a BdeDatabase.pm object method and you don't need to make all of the lower level dbh calls.

Reply:
What's the difference between a new API and promotion of existing to public ?
This was a direct port of the stored procedure by the same name so didn't want to change
anything but the COPY line (to also help with comparing).

Reply:
private method, those starting with a underscore are only used internally within the PgSQL stored procedures and their API signature can be changed without wider review outside the PgSQL code. The Public API are main workflow calls used by external software e.g linz_bde_uploader. I think the best think to do here is review if private calls that you have ported into linz_bde_uploader can be replaced with public calls and the private functions can be promoted to public

Reply:
I see we have 3 levels here: Stored procedure (can have their private), Perl library (can have their private) and Scripts (linz_bde_uploader). This implementation is having the Perl library use stored-procedure-private, keeping the external interface untouched (I think, or maybe I should somehow tag/rename uploadDataToTempTable to make it effectively module-private, not sure).

I'll first profile the difference between the two COPY implementation and after that see how avoiding perl modules using private pg-schema functions can be avoided.

Reply:
Good approach.